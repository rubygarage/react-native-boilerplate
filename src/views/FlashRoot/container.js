import React, { Component } from 'react';
import { connect } from 'react-redux';
import { UIManager, LayoutAnimation } from 'react-native';
import PropTypes from 'prop-types';

import { flashesSelector } from 'state/flash/selectors';
import FlashRootComponent from './component';

const ANIMATION_DURATION = 250;

class FlashRoot extends Component {
  componentDidMount() {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }

  componentDidUpdate(prevProps) {
    const { flashes: prevFlashes } = prevProps;
    const { flashes } = this.props;
    if (prevFlashes.length < flashes.length) {
      this.doAnimation();
    }
  }

  doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  render() {
    return (
      <FlashRootComponent
        {...this.props}
        doAnimation={this.doAnimation}
      />
    );
  }
}

FlashRoot.propTypes = {
  flashes: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

const mapStateToProps = (state) => ({
  flashes: flashesSelector(state),
});

export { FlashRoot as FlashRootContainer };
export default connect(mapStateToProps)(FlashRoot);
