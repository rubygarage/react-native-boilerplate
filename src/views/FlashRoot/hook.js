import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { UIManager, LayoutAnimation } from 'react-native';

import { flashesSelector } from 'state/flash/selectors';

export const ANIMATION_DURATION = 250;

const useContainer = () => {
  const flashes = useSelector(flashesSelector);

  const doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  useEffect(() => {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(doAnimation, [flashes.length]);

  return {
    doAnimation,
  };
};

export default useContainer;
