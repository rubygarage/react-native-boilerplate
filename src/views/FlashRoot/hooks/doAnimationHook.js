import { LayoutAnimation } from 'react-native';

export const ANIMATION_DURATION = 250;

const doAnimationHook = () => {
  const doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  return {
    doAnimation,
  };
};

export default doAnimationHook;
