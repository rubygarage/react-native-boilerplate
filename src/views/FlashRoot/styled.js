import styled from 'styled-components/native';

const List = styled.FlatList`
  position: absolute;
  top: ${({ topInset }) => topInset}px;
  width: 100%;
  margin-top: 12px;
  margin-bottom: 12px;
  padding-left: 12px;
  padding-right: 12px;
  z-index: 5;
`;

// eslint-disable-next-line import/prefer-default-export
export { List };
