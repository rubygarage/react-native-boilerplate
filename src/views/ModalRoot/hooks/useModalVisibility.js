import { useState } from 'react';
import { useDispatch } from 'react-redux';

import { hideModal } from 'state/modal/actions';

const useModalVisibility = () => {
  const dispatch = useDispatch();

  const [isVisible, setIsVisible] = useState(true);

  const onCloseModal = () => {
    setIsVisible(false);
  };

  const onDestroyModal = () => {
    dispatch(hideModal());
    setIsVisible(true);
  };

  return {
    isVisible,
    setIsVisible,
    onCloseModal,
    onDestroyModal,
  };
};

export default useModalVisibility;
