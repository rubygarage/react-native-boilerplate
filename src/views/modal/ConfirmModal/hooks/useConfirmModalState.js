import { useState } from 'react';
import { useDispatch } from 'react-redux';

const useConfirmModalState = (props) => {
  const { onCloseModal, onDestroyModal, confirmAction } = props;

  const dispatch = useDispatch();

  const [isShouldConfirm, setShouldConfirm] = useState(false);

  const handleShouldConfirm = () => {
    setShouldConfirm(true);
    onCloseModal();
  };

  const onModalHide = () => {
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  };

  return {
    handleShouldConfirm,
    onModalHide,
    setShouldConfirm,
    isShouldConfirm,
  };
};

export default useConfirmModalState;
