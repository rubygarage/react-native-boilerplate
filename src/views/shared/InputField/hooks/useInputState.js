import { useState } from 'react';
import { call, path, pathOr } from 'ramda';

function useInputState(props) {
  const {
    field: { name },
    form: { handleBlur, errors },
  } = props;

  const [isFocused, setIsFocused] = useState(false);
  const [actionWidth, setActionWidth] = useState(0);

  /**
   * On Action required.
   */
  const onActionRendered = (event) => {
    const { width } = event.nativeEvent.layout;

    setActionWidth(width);
  };

  /**
   * On focus handler.
   */
  const onFocus = () => {
    setIsFocused(true);
  };

  /**
   * On Blur handler
   */
  const onBlur = (event) => {
    call(handleBlur(name), event);
    setIsFocused(false);
  };

  /**
   * Error data
   */
  const getErrorData = () => {
    const errorId = path([name, 'id'], errors);
    const errorValues = path([name, 'values'], errors);
    const errorIcon = pathOr('alert', [name, 'icon'], errors);
    return { errorId, errorValues, errorIcon };
  };

  return {
    isFocused,
    actionWidth,
    onActionRendered,
    onFocus,
    onBlur,
    getErrorData,
    setIsFocused,
    setActionWidth,
  };
}

export default useInputState;
