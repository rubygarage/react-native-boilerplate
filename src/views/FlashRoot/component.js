import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { UIManager } from 'react-native';

import { flashesSelector } from 'state/flash/selectors';

import Flash from './Flash';
import * as S from './styled';
import doAnimationHook from './hooks/doAnimationHook';

const keyExtractor = (item) => item.id;

const FlashRootComponent = ({ topInset }) => {
  const flashes = useSelector(flashesSelector);

  const { doAnimation } = doAnimationHook();

  useEffect(() => {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(() => {
    doAnimation();

    // eslint-disable-next-line
  }, [flashes.length]);

  return (
    <S.List
      data={flashes}
      renderItem={({ item }) => <Flash {...item} doAnimation={doAnimation} />}
      keyExtractor={keyExtractor}
      showsVerticalScrollIndicator={false}
      topInset={topInset}
    />
  );
};

FlashRootComponent.propTypes = {
  topInset: PropTypes.number,
};

FlashRootComponent.defaultProps = {
  topInset: 0,
};

export default FlashRootComponent;
