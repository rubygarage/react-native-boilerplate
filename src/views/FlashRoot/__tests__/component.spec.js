import React from 'react';
import { shallow } from 'enzyme';
import { useSelector } from 'react-redux';
import { UIManager } from 'react-native';

import { ALERT_TYPES } from 'constants/alert';
import FlashRoot from '../component';
import * as doAnimationHook from '../hooks/doAnimationHook';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');

  RN.UIManager.setLayoutAnimationEnabledExperimental = jest.fn();
  RN.LayoutAnimation.configureNext = jest.fn();

  return RN;
});

const flashes = [
  { id: '1', flashProps: {} },
  { id: '2', flashProps: {} },
];

describe('FlashRoot component', () => {
  const flash = { id: '0', type: ALERT_TYPES.info, title: 'test_title', message: 'test_message' };
  const defaultProps = {
    flashes: [flash],
    doAnimation: jest.fn(),
    flashProps: {},
  };
  let component = null;
  const doAnimation = jest.fn();

  doAnimationHook.default = () => ({
    doAnimation,
  });

  beforeEach(() => {
    useSelector.mockImplementation(() => flashes);

    component = shallow(<FlashRoot {...defaultProps} />);
  });

  it('renders correctly by default', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correct item', () => {
    const flatList = component.find('Styled(FlatList)');

    const item = flatList.renderProp('renderItem')({ item: flash });
    expect(item).toMatchSnapshot();
  });

  it('returns correct key', () => {
    const sectionList = component.find('Styled(FlatList)');

    const key = sectionList.prop('keyExtractor')(flash);
    expect(key).toEqual(flash.id);
  });

  describe('Lifecycle', () => {
    it('useEffect - mounting, should call setLayoutAnimationEnabledExperimental', () => {
      shallow(<FlashRoot {...defaultProps} />);

      expect(UIManager.setLayoutAnimationEnabledExperimental).toHaveBeenCalledWith(true);
    });

    it('useEffect - mounting, should not call setLayoutAnimationEnabledExperimental', () => {
      UIManager.setLayoutAnimationEnabledExperimental = null;

      shallow(<FlashRoot {...defaultProps} />);

      expect(UIManager.setLayoutAnimationEnabledExperimental).toBeFalsy();
    });

    it('useEffect - updating', () => {
      useSelector.mockImplementation(() => [...flashes, { id: '3' }]);

      // Force update. It is only way to update component with useEffect calling.
      component.setProps();

      expect(doAnimation).toHaveBeenCalled();
    });
  });
});
