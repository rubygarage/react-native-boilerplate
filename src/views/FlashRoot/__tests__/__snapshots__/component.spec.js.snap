// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashRoot component renders correct item 1`] = `
<Flash
  doAnimation={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  flashProps={Object {}}
  id="0"
  message="test_message"
  title="test_title"
  type="info"
/>
`;

exports[`FlashRoot component renders correctly by default 1`] = `
<Styled(FlatList)
  data={
    Array [
      Object {
        "flashProps": Object {},
        "id": "1",
      },
      Object {
        "flashProps": Object {},
        "id": "2",
      },
    ]
  }
  keyExtractor={[Function]}
  renderItem={[Function]}
  showsVerticalScrollIndicator={false}
  topInset={0}
/>
`;
